File: scripts/report.py
#!/usr/bin/env python3
import html
import json
import os
import re
import subprocess
import sys
import urllib.request


def compute_sizes(node, parent_path):
    total_size = 0
    if "contents" in node:
        for child in node["contents"]:
            child_size = compute_sizes(child, os.path.join(parent_path, node["name"]))
            total_size += child_size
        node["size"] = total_size
    else:
        file_path = os.path.join(parent_path, node["name"])
        total_size = os.path.getsize(file_path) if os.path.isfile(file_path) else 0
        node["size"] = total_size
    return total_size


def sort_by_size(node):
    if "contents" in node:
        node["contents"].sort(key=lambda x: x["size"], reverse=True)
        for child in node["contents"]:
            sort_by_size(child)


def build_html(node, parent_path):
    name = html.escape(node["name"])
    if "contents" in node:
        size_mb = f"{(node['size']/1024/1024):.2f} MB"
        s = f"<details><summary>{name} ({size_mb})</summary>\n"
        for child in node["contents"]:
            s += build_html(child, os.path.join(parent_path, node["name"]))
        s += "</details>\n"
        return s
    else:
        size_mb = f"{(node['size']/1024/1024):.2f} MB"
        rel_path = html.escape(os.path.join(parent_path, node["name"]))
        return f'<div><a href="{rel_path}">{name}</a> ({size_mb})</div>\n'


def get_os_version():
    try:
        with open("/etc/os-release", "r", encoding="utf-8") as f:
            for line in f:
                if line.startswith("PRETTY_NAME="):
                    return line.split("=", 1)[1].strip().strip('"')
        return "Unknown OS"
    except FileNotFoundError:
        return "Unknown OS"


def get_version(command, args=["--version"]):
    try:
        result = subprocess.run(
            [command] + args,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )
        return result.stdout.strip()
    except FileNotFoundError:
        return "not installed"


def extract_body(html_text):
    start = html_text.find("<body>")
    end = html_text.find("</body>")
    if start != -1 and end != -1:
        return html_text[start + len("<body>") : end]
    return html_text


def fetch_previous_report():
    url = "https://hilll.dev/report"
    try:
        with urllib.request.urlopen(url) as response:
            remote_html = response.read().decode("utf-8")
    except Exception:
        return None

    pattern = re.compile(r"<!-- REPORT START -->(.*?)<!-- REPORT END -->", re.DOTALL)
    reports = pattern.findall(remote_html)
    if not reports:
        body = extract_body(remote_html)
        return body if body.strip() else None
    return reports[-2] if len(reports) >= 2 else reports[-1]


def main():
    # ---- Capture build time from CLI args (if provided) ----
    build_time = 0
    if len(sys.argv) > 1:
        try:
            build_time = int(sys.argv[1])
        except ValueError:
            print("Invalid build time argument. Ignoring.")
    else:
        print("No build time provided")

    # ---- Gather version info ----
    time_str = get_version("date", ["+%x (%A) %X %z"])
    os_version = get_os_version()
    nix_version = get_version("nix")
    quartz_version = get_version("npx", ["quartz", "--version"]).split("\n").pop()
    python_version = get_version("python")
    core_pack = get_version("corepack")
    node_version = get_version("node")
    npm_version = get_version("npm")
    pnpm_version = get_version("pnpm")
    cargo_version = get_version("cargo")
    rustc_version = get_version("rustc")
    rustup_version = get_version("rustup").split("\n").pop(0)
    wasm_pack_version = get_version("wasm-pack")
    tsc_version = get_version("pnpm", ["exec", "tsc", "--version"])
    # Retrieve git commit hash
    git_commit = get_version("git", ["rev-parse", "HEAD"])
    os.chdir("../ts")
    webpack_full = get_version("pnpm", ["exec", "webpack", "--version"])
    webpack_version = (
        webpack_full.split("Packages:")[1].replace("\n", "<br>")
        if "Packages:" in webpack_full
        else webpack_full
    )
    os.chdir("../public")

    # ---- Run 'tree' command to gather folder structure info ----
    folder_to_scan = "."
    result = subprocess.run(
        ["tree", "-J", folder_to_scan], stdout=subprocess.PIPE, text=True
    )
    data = json.loads(result.stdout)
    report_info = None

    # ---- Compute sizes and sort ----
    for top_node in data:
        if top_node["type"] == "report":
            report_info = top_node
        else:
            compute_sizes(top_node, "")
            sort_by_size(top_node)

    # ---- Build current report HTML with markers ----
    current_body = []
    current_body.append("<h1>Build Report</h1>")
    current_body.append(f"<p><strong>Report generated at:</strong> {time_str}</p>")
    if build_time > 0:
        current_body.append(f"<p><strong>Build Time:</strong> {build_time} seconds</p>")
    current_body.append(f"<p><strong>Operating System:</strong> {os_version}</p>")
    current_body.append(f"<p><strong>Nix:</strong> {nix_version}</p>")
    current_body.append(f"<p><strong>Quartz:</strong> {quartz_version}</p>")
    current_body.append(f"<p><strong>Python:</strong> {python_version}</p>")
    current_body.append(f"<p><strong>corepack:</strong> {core_pack}</p>")
    current_body.append(f"<p><strong>Node.js:</strong> {node_version}</p>")
    current_body.append(f"<p><strong>npm:</strong> {npm_version}</p>")
    current_body.append(f"<p><strong>pnpm:</strong> {pnpm_version}</p>")
    current_body.append(f"<p><strong>cargo:</strong> {cargo_version}</p>")
    current_body.append(f"<p><strong>rustc:</strong> {rustc_version}</p>")
    current_body.append(f"<p><strong>rustup:</strong> {rustup_version}</p>")
    current_body.append(f"<p><strong>wasm-pack:</strong> {wasm_pack_version}</p>")
    current_body.append(f"<p><strong>tsc:</strong> {tsc_version}</p>")
    current_body.append(f"<p><strong>webpack:</strong> {webpack_version}</p>")
    # Add git commit hash
    current_body.append(f"<p><strong>Git Commit:</strong> {git_commit}</p>")
    if report_info is not None:
        directories = report_info.get("directories", 0)
        files = report_info.get("files", 0)
        current_body.append(
            f"<h2>Tree Summary</h2><p>Total Directories: {directories}, Total Files: {files}</p>"
        )
    current_body.append("<h2>Folder Contents</h2>")
    for top_node in data:
        if top_node["type"] != "report":
            current_body.append(build_html(top_node, ""))
    current_report_body = "\n".join(current_body)
    current_report = (
        f"<!-- REPORT START -->\n{current_report_body}\n<!-- REPORT END -->"
    )

    # ---- Fetch previous report (if available) ----
    previous_report = fetch_previous_report()

    # ---- Build combined HTML with side-by-side columns ----
    combined_html = f"""<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8"/>
  <title>Build Report Comparison</title>
  <style>
    .container {{
      display: flex;
      flex-wrap: wrap;
    }}
    .report-column {{
      flex: 1;
      padding: 10px;
      box-sizing: border-box;
      min-width: 300px;
      border: 1px solid #ccc;
      margin: 5px;
    }}
  </style>
</head>
<body>
  <h1>Build Report Comparison</h1>
  <div class="container">
    <div class="report-column"><h2>Current Report</h2>{current_report}</div>
    {"<div class='report-column'><h2>Previous Report</h2>" + previous_report + "</div>" if previous_report else ""}
  </div>
</body>
</html>
"""

    # ---- Write the combined report to 'report.html' ----
    with open("report.html", "w", encoding="utf-8") as f:
        f.write(combined_html)


if __name__ == "__main__":
    main()

File: tools/site/src/report.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use crate::{os_args, Result, Site};

const REPORT_START: &str = "<!-- REPORT START -->";
const REPORT_END: &str = "<!-- REPORT END -->";

#[derive(Default)]
struct TreeSummary {
    directories: usize,
    files: usize,
}

struct TreeNode {
    name: String,
    rel_path: PathBuf,
    size: u64,
    is_dir: bool,
    children: Vec<TreeNode>,
}

pub(crate) fn write(site: &Site, public: &Path, build_time: u64) -> Result<()> {
    let time_str = command_text(public, "date", &["+%x (%A) %X %z"]);
    let os_version = os_version();
    let nix_version = command_text(public, "nix", &["--version"]);
    let quartz_version = command_text(&site.root, "npx", &["quartz", "--version"])
        .lines()
        .last()
        .unwrap_or("not installed")
        .to_string();
    let python_version = command_text(public, "python", &["--version"]);
    let corepack_version = command_text(public, "corepack", &["--version"]);
    let node_version = command_text(public, "node", &["--version"]);
    let npm_version = command_text(public, "npm", &["--version"]);
    let pnpm_version = command_text(public, "pnpm", &["--version"]);
    let cargo_version = command_text(public, "cargo", &["--version"]);
    let rustc_version = command_text(public, "rustc", &["--version"]);
    let rustup_version = command_text(public, "rustup", &["--version"])
        .lines()
        .next()
        .unwrap_or("not installed")
        .to_string();
    let wasm_pack_version = command_text(public, "wasm-pack", &["--version"]);
    let tsc_version = command_text(public, "pnpm", &["exec", "tsc", "--version"]);
    let git_commit = site
        .output_optional(&site.root, "git", &os_args(&["rev-parse", "HEAD"]))?
        .unwrap_or_else(|| "unknown".to_string());
    let webpack_version = webpack_version(&site.root.join("ts"));

    let (tree, summary) = scan_public(public)?;
    let previous_report = fetch_previous_report();

    let mut current_body = Vec::new();
    current_body.push("<h1>Build Report</h1>".to_string());
    current_body.push(format!(
        "<p><strong>Report generated at:</strong> {}</p>",
        escape_html(&time_str)
    ));
    if build_time > 0 {
        current_body.push(format!(
            "<p><strong>Build Time:</strong> {build_time} seconds</p>"
        ));
    }
    current_body.push(format!(
        "<p><strong>Operating System:</strong> {}</p>",
        escape_html(&os_version)
    ));
    current_body.push(version_line("Nix", &nix_version));
    current_body.push(version_line("Quartz", &quartz_version));
    current_body.push(version_line("Python", &python_version));
    current_body.push(version_line("corepack", &corepack_version));
    current_body.push(version_line("Node.js", &node_version));
    current_body.push(version_line("npm", &npm_version));
    current_body.push(version_line("pnpm", &pnpm_version));
    current_body.push(version_line("cargo", &cargo_version));
    current_body.push(version_line("rustc", &rustc_version));
    current_body.push(version_line("rustup", &rustup_version));
    current_body.push(version_line("wasm-pack", &wasm_pack_version));
    current_body.push(version_line("tsc", &tsc_version));
    current_body.push(version_line("webpack", &webpack_version));
    current_body.push(version_line("Git Commit", &git_commit));
    current_body.push(format!(
        "<h2>Tree Summary</h2><p>Total Directories: {}, Total Files: {}</p>",
        summary.directories, summary.files
    ));
    current_body.push("<h2>Folder Contents</h2>".to_string());
    current_body.push(build_tree_html(&tree));

    let current_report_body = current_body.join("\n");
    let current_report = format!("{REPORT_START}\n{current_report_body}\n{REPORT_END}");

    let previous_column = previous_report
        .map(|report| format!("<div class='report-column'><h2>Previous Report</h2>{report}</div>"))
        .unwrap_or_default();

    let combined_html = format!(
        r#"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8"/>
  <title>Build Report Comparison</title>
  <style>
    .container {{
      display: flex;
      flex-wrap: wrap;
    }}
    .report-column {{
      flex: 1;
      padding: 10px;
      box-sizing: border-box;
      min-width: 300px;
      border: 1px solid #ccc;
      margin: 5px;
    }}
  </style>
</head>
<body>
  <h1>Build Report Comparison</h1>
  <div class="container">
    <div class="report-column"><h2>Current Report</h2>{current_report}</div>
    {previous_column}
  </div>
</body>
</html>
"#
    );

    fs::write(public.join("report.html"), combined_html)?;
    Ok(())
}

fn version_line(label: &str, value: &str) -> String {
    format!(
        "<p><strong>{}:</strong> {}</p>",
        escape_html(label),
        escape_html(value)
    )
}

fn command_text(cwd: &Path, program: &str, args: &[&str]) -> String {
    let output = Command::new(program)
        .args(args)
        .current_dir(cwd)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output();

    match output {
        Ok(output) => {
            let mut text = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if text.is_empty() {
                text = String::from_utf8_lossy(&output.stderr).trim().to_string();
            }
            if text.is_empty() {
                "not installed".to_string()
            } else {
                text
            }
        }
        Err(_) => "not installed".to_string(),
    }
}

fn webpack_version(ts_dir: &Path) -> String {
    let full = command_text(ts_dir, "pnpm", &["exec", "webpack", "--version"]);
    if let Some((_, packages)) = full.split_once("Packages:") {
        packages.trim().replace('\n', "<br>")
    } else {
        full
    }
}

fn os_version() -> String {
    let Ok(source) = fs::read_to_string("/etc/os-release") else {
        return "Unknown OS".to_string();
    };

    source
        .lines()
        .find_map(|line| {
            line.strip_prefix("PRETTY_NAME=")
                .map(|value| value.trim_matches('"').to_string())
        })
        .unwrap_or_else(|| "Unknown OS".to_string())
}

fn scan_public(public: &Path) -> Result<(TreeNode, TreeSummary)> {
    let mut summary = TreeSummary::default();
    let mut root = scan_node(public, PathBuf::from("."), &mut summary)?;
    sort_tree(&mut root);
    Ok((root, summary))
}

fn scan_node(path: &Path, rel_path: PathBuf, summary: &mut TreeSummary) -> Result<TreeNode> {
    let name = rel_path
        .file_name()
        .map(|name| name.to_string_lossy().into_owned())
        .unwrap_or_else(|| ".".to_string());

    if path.is_dir() {
        let mut children = Vec::new();
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let child_path = entry.path();
            let child_name = entry.file_name();
            let child_rel = if rel_path.as_os_str() == "." {
                PathBuf::from(child_name)
            } else {
                rel_path.join(child_name)
            };

            if child_path.is_dir() {
                summary.directories += 1;
            } else {
                summary.files += 1;
            }
            children.push(scan_node(&child_path, child_rel, summary)?);
        }

        let size = children.iter().map(|child| child.size).sum();
        Ok(TreeNode {
            name,
            rel_path,
            size,
            is_dir: true,
            children,
        })
    } else {
        Ok(TreeNode {
            name,
            rel_path,
            size: fs::metadata(path)
                .map(|metadata| metadata.len())
                .unwrap_or(0),
            is_dir: false,
            children: Vec::new(),
        })
    }
}

fn sort_tree(node: &mut TreeNode) {
    node.children.sort_by(|left, right| {
        right
            .size
            .cmp(&left.size)
            .then_with(|| left.name.cmp(&right.name))
    });
    for child in &mut node.children {
        sort_tree(child);
    }
}

fn build_tree_html(node: &TreeNode) -> String {
    let name = escape_html(&node.name);
    let size = format!("{:.2} MB", node.size as f64 / 1024.0 / 1024.0);

    if node.is_dir {
        let mut html = format!("<details><summary>{name} ({size})</summary>\n");
        for child in &node.children {
            html.push_str(&build_tree_html(child));
        }
        html.push_str("</details>\n");
        html
    } else {
        let rel_path = escape_html(&path_for_html(&node.rel_path));
        format!(r#"<div><a href="{rel_path}">{name}</a> ({size})</div>"#) + "\n"
    }
}

fn path_for_html(path: &Path) -> String {
    path.to_string_lossy().replace('\\', "/")
}

fn fetch_previous_report() -> Option<String> {
    let output = Command::new("curl")
        .args(["-fsSL", "--max-time", "10", "https://hilll.dev/report"])
        .stderr(Stdio::null())
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let remote_html = String::from_utf8_lossy(&output.stdout);
    let reports = extract_marked_reports(&remote_html);
    if reports.len() >= 2 {
        reports.get(reports.len() - 2).cloned()
    } else if let Some(report) = reports.last() {
        Some(report.clone())
    } else {
        extract_body(&remote_html).filter(|body| !body.trim().is_empty())
    }
}

fn extract_marked_reports(html: &str) -> Vec<String> {
    let mut reports = Vec::new();
    let mut offset = 0;

    while let Some(start) = html[offset..].find(REPORT_START) {
        let body_start = offset + start + REPORT_START.len();
        let Some(end) = html[body_start..].find(REPORT_END) else {
            break;
        };
        let body_end = body_start + end;
        reports.push(html[body_start..body_end].to_string());
        offset = body_end + REPORT_END.len();
    }

    reports
}

fn extract_body(html: &str) -> Option<String> {
    let start = html.find("<body>")?;
    let end = html.find("</body>")?;
    Some(html[start + "<body>".len()..end].to_string())
}

fn escape_html(value: &str) -> String {
    let mut escaped = String::with_capacity(value.len());
    for ch in value.chars() {
        match ch {
            '&' => escaped.push_str("&amp;"),
            '<' => escaped.push_str("&lt;"),
            '>' => escaped.push_str("&gt;"),
            '"' => escaped.push_str("&quot;"),
            '\'' => escaped.push_str("&#39;"),
            _ => escaped.push(ch),
        }
    }
    escaped
}
